File: exercises/week02/src/solid/ControlUnit.java
package solid;

/**
 * The main entry point for controlling sensors
 */
public class ControlUnit {
    /**
     * Create some sensors, poll them every 5 seconds, and report any that are triggered (three responsibilities).
     */
    private void pollSensors() {
        Sensor sensors[] = {
                new FireSensor("Lobby"),
                new SmokeSensor("Stairs"),
                new SmokeSensor("Mezzanine"),
                new FireSensor("Kitchen")
        };

        while (true) {
            for (Sensor sensor : sensors) {
                if (sensor.isTriggered()) {
                    System.out.println("Alert! " + sensor.getSensorType() + " in " + sensor.getLocation() + " triggered");
                }
                if (sensor.getBatteryPercentage() < 5) {
                    System.out.println("Alert! " + sensor.getSensorType() +
                            " in " + sensor.getLocation() + " battery critical");

                } else if (sensor.getBatteryPercentage() < 15) {
                    System.out.println("Warning: " + sensor.getSensorType() +
                            " in " + sensor.getLocation() + " battery low");
                }
            }

            try {
                Thread.sleep(5000);
            } catch (InterruptedException e) {
                // If sleep is interrupted, we simply check more frequently, so do nothing here.
            }
        }
    }

    public static void main(String[] args) {
        ControlUnit controlUnit = new ControlUnit();
        controlUnit.pollSensors();
    }
}

File: exercises/week02/src/solid/ControlUnit.scala
object ControlUnit

/**
  * The primary constructor of ControlUnit is used to inject the dependency, namely, the list of sensors.
  * @param sensors
  */
class ControlUnit(sensors: Traversable[Sensor]) {

  /**
    * This method checks whether sensors are triggered, and displays messages reporting their state.
    */
  def pollSensors() {
    for (sensor <- sensors) {
      if (sensor.isTriggered) {
        System.out.println("A " + sensor.getSensorType + " sensor was triggered at " + sensor.getLocation)
      }
      else {
        System.out.println("Polled " + sensor.getSensorType + " at " + sensor.getLocation + " successfully")
      }
    }
  }
}
